src/eval/lang.rs

```diff
@@ -163,33 +152,40 @@ impl<F: LurkField, C: Coprocessor<F>> Lang<F, C> {
         maybe_sym.and_then(|sym| self.coprocessors.get(&sym))
     }
 
+    #[inline]
     pub fn has_coprocessors(&self) -> bool {
         !self.coprocessors.is_empty()
     }
 
+    #[inline]
     pub fn is_default(&self) -> bool {
         !self.has_coprocessors()
     }
 
+    #[inline]
     pub fn coprocessor_count(&self) -> usize {
-        self.index.len()
+        self.coprocessors.len()
     }
 
     pub fn get_index(&self, z_ptr: &ZExprPtr<F>) -> Option<usize> {
-        self.index.get_index_of(z_ptr)
+        for (i, (_, z_ptr_lang)) in self.coprocessors.values().enumerate() {
+            if z_ptr == z_ptr_lang {
+                return Some(i);
+            }
+        }
```

Comment:
I realize that `get_index_of` has the same property, but this replacement emphasizes that we should avoid a scan. I would rather see that problem solved — than strict minimization of the number of data structures used, when we refactor. What you see now exists simply to define the necessary interface so NIVC coprocessors can work. We should indeed refactor to provide this more simply, but we should also address the 'complexity' so the needed operations are also cheap (i.e. O(1) vs O(n)).

We need to be able to look things up by symbol, by index, and by z_ptr. This probably means we need to maintain a second index. Most simply, that could be a map from z_ptr to index, to solve the problem here.

I'm fine with simplifying, but I would be happiest if we also solved this problem at the same time, since that will be most stable and reduce the likelihood that we forget to come back to this.

Reply:
I think the O(1) vs O(n) here is reasonably affected by the overhead of using a data structure that uses hashing, since we don't expect the number of coprocessors to grow too large.

I can keep the `IndexSet<ZExprPtr<F>>` if you want, but the benefit is not so clear to me in this case.

Reply:
I'm not claiming there's a benefit to one bad way vs another. I'm asking that we optimize this function. If you have time to do it once here and now, that would be the simplest/best way.

Your statement that 'we don't expect the number of coprocessors to grow too large' carries an assumption I would not like baked into design discussions. Quite the opposite, which is part of why I'd like to make this a non-issue.